### PR TITLE
Validate all xDS resources before sending them to Envoy

### PR DESCRIPTION
**Pull request by rlenglet**
Relates to: https://github.com/istio/istio/issues/6099
Signed-off-by: Romain Lenglet <romain@covalent.io>

### CODE REVIEW

**Comment by rlenglet** [comment]
cc @kyessenov

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -574,6 +574,16 @@ func (s *DiscoveryServer) pushRoute(con *XdsConnection) error {
...
 		}
 
+		if err = r.Validate(); err != nil {
+			retErr := fmt.Errorf("RDS: Generated invalid route %s for node %s: %v", routeName, con.modelNode, err)
+			adsLog.Errorf("RDS: Generated invalid routes for route %s for node %s: %v, %v", routeName, con.modelNode, err, r)
+			pushes.With(prometheus.Labels{"type": "rds_builderr"}).Add(1)
+			// Generating invalid routes is a bug.
+			// Panic instead of trying to recover from that, since we can't
+			// assume anything about the state.
+			panic(retErr.Error())
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/ads.go]: Did you encounter an invalid route from Pilot ?

**Comment by rshriram** [comment]
Please make sure cloud foundry v1alpha3-v2 and envoyv2-v1alpha3 tests pass.

**Comment by rlenglet** [comment]
> Just curious. Did you see invalid config from Pilot? I plugged several sources of invalid config, choosing to not emit the listener/route/cluster rather than emitting an invalid one.

@rshriram Yes, I saw an invalid listener from Pilot, cf. https://github.com/istio/istio/issues/6099

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 1 above)
Comment (rlenglet) [reply pilot/pkg/proxy/envoy/v2/ads.go]: No, I just saw an invalid listener. I'm just making sure here that we'll never send any invalid resource of any of the currently supported types.

**Comment by rshriram** [comment]
can you sync your branch with master?

**Comment by ostromart** [comment]
This will definitely surface any lingering problems with invalid configs and force us to deal with them, so in general I'm supportive. Do we need to pre-validate in pilot, since in general envoy will tell us if a config is bad?

**Comment by rshriram** [comment]
It might crash or reject config. But its better to know about the rejection in Pilot and deal with it before hand.

**Comment by kyessenov** [comment]
Do keep in mind that envoy can reject config that passes validation. Not all checks are static (e.g. parsing protos from structs).

**Comment by rlenglet** [comment]
> Do we need to pre-validate in pilot, since in general envoy will tell us if a config is bad?

Yes, since Pilot's xDS implementation doesn't seem to deal properly with NACKs from Envoy.

**Pull request merged**